### PR TITLE
HDDS-9844. [hsync] De-synchronize write APIs (v3)

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

Same as https://github.com/apache/ozone/pull/6413 but a third impl based on master branch. FYI. NOT ready for review.

This is a much cleaner impl than v2 (#6413) after a bit more understanding of what is going on in the client with buffer/future stuff. But the tests won't pass at this point.

cc @duongkame 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9844

## How was this patch tested?

- See #6413

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/AbstractCommitWatcher.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ abstract class AbstractCommitWatcher<BUFFER> {
 
   private final XceiverClientSpi client;
 
-  private long totalAckDataLength;
+  private volatile long totalAckDataLength;
```
Comment (jojochuang) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/AbstractCommitWatcher.java:65]: consider using AtomicLong instead. It supports addAndGet() which is exactly what we need.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java [hunk 2]
```diff
@@ -205,8 +209,10 @@ public synchronized void write(byte[] b, int off, int len)
     if (len == 0) {
       return;
     }
-    handleWrite(b, off, len, false);
-    writeOffset += len;
+    synchronized (this) {
+      handleWrite(b, off, len, false);
```
Comment (jojochuang) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java:213]: handleWrite() calls allocateBlockIfNeeded() which means it's going to get blocked waiting for OM response, which is fine. It doesn't happen frequent so shouldn't be a big deal. We can revisit later if it turns out to be not the case.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java [hunk 3]
```diff
@@ -267,50 +266,115 @@ private XceiverClientReply newWatchReply(
...
     }
 
+//    return watchForCommitInternalOld(index);
+    return watchForCommitInternalNew(index);
+  }
+
+  /**
+   * Old implementation that handles the retry using try-catch with a new future.
+   */
+  private XceiverClientReply watchForCommitInternalOld(long index)
```
Comment (jojochuang) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:282]: dead code.

Comment (smengcl) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/XceiverClientRatis.java:282]: This is intentionally left there for debugging. This PR is not intended for review atm.

**Review by duongkame — COMMENTED** [review]
Just want to put down what was discussed offline. 

The logic of `hsync()` consists of the following.
1. Update the current buffer and internal counters in the OutputStream states.
2. Use the current buffer and internal counters to compose a `writeChunk` request.
3. Send the `writeChunk` request.
4. Wait the `writeChunk` response and send a putBlock request.
5. Wait for `putBlock` reply and issue a watchForCommmit.
6. Wait for `watchForCommit` response and release retries buffers accordingly (in CommitWatcher).

I strongly think the step 1, 2 and 3 should be in synchronized block leverage the calling thread.
Step 4, 5, 6 can be done asynchronously. Please note the current code is done on the assumption that the OutputStream is not thread-safe, so the current impl of 4 5 6 relies on state/counters in the OutputStream instances (e.g. commit to watch, index of the buffer to clear...). This becomes obsolete when having simultanous hsync(). We need to access each step logic and identify the data needed for each calls aand make the required refactoring.

A psuedo code of `hsync()` would looks like.

```
void hsync() {
	
	CompleteableFuture<WriteChunkResponse> writeChunkResp;
	synchronize(this) {
		// 1. Update the current buffer and internal counters in the OutputStream states.
		// 2. Use the current buffer and internal counters to compose a `writeChunk` request.
		writeChunk = getWriteChunkRequest();
		// 3. Send the `writeChunk` request.
		CompleteableFuture<WriteChunkResponse> writeChunkResp = sendWriteChunk(writeChunk);
	}
	// 4. Wait the `writeChunk` response and send a putBlock request.
	CompleteableFuture<PutBlockResponse> putBlockReps = writeChunkResp.thenApply(x -> executePutBlock(x));
	// 5. Wait for `putBlock` reply and issue a watchForCommmit.
	CompleteableFuture<WatchForCommitResponse> watchForReps = writeChunkResp.thenApply(x -> watchForCommit(x));
	// 6. Wait for `watchForCommit` response and release retries buffers accordingly (in CommitWatcher).
	WatchForCommitResponse watchForCommit = watchForReps.get();
}
```

Similar things can be done for `write()`.

I know it's easy to say, but I believe the change require a great amount of refactor to get desyncrhonize hsync() and write().

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/KeyOutputStream.java:213]: Yes that is correct.

**Comment by smengcl** [comment]
> Just want to put down what was discussed offline.
> 
> The logic of `hsync()` consists of the following.
> 
> 1. Update the current buffer and internal counters in the OutputStream states.
> 2. Use the current buffer and internal counters to compose a `writeChunk` request.
> 3. Send the `writeChunk` request.
> 4. Wait the `writeChunk` response and send a putBlock request.
> 5. Wait for `putBlock` reply and issue a watchForCommmit.
> 6. Wait for `watchForCommit` response and release retries buffers accordingly (in CommitWatcher).

Thanks @duongkame the comment. Overall I agree with your idea.

Commenting on some details:

1. In step 4, if I understand correctly with the current master branch impl, the writeChunk response is NOT actually waited on before sending out the putBlock request. BOS#writeChunkToContainer() returns a future but the caller BOS#writeChunk() doesn't even care.
2. In step 5, watchForCommit itself waits for all pending unacknowledged putBlock requests to finish, not before.

Effectively it is already sending writeChunk and putBlock in parallel, so we can further improve the parallelism.

**Comment by duongkame** [comment]
> 
> Commenting on some details:
> 
> 1. In step 4, if I understand correctly with the current master branch impl, the writeChunk response is NOT actually waited on before sending out the putBlock request. BOS#writeChunkToContainer() returns a future but the caller BOS#writeChunk() doesn't even care.
> 2. In step 5, watchForCommit itself waits for all pending unacknowledged putBlock requests to finish, not before.
> 
> Effectively it is already sending writeChunk and putBlock in parallel, so we can further improve the parallelism.

That's correct, @smengcl. I just found out how PutBlock data is done. The field `BOS.containerBlockData` accumulates all the `writeChunk` requests that have been sent by the BOS, and that data field is sent with putBlock to confirm with datanodes about the sent chunks.
So, the putBlock can be sent without waiting for writeChunk, but it must be sent *after* (to ensure the order of sending requests to Ratis) and contain the previous writeChunk in its data. We don't event need to wait for `writeChunk` (as the current implementation) because the `watchForCommit` will wait for the `PutBlock` to get fully replicated (and putBlock is sent after writeChunk, so when putBlock is replicated, writeChunk sure is).

```
void hsync() {
	
	CompleteableFuture<PutBlockRespose> putBlockResponse;
	synchronize(this) {
		// 1. Update the current buffer and internal counters in the OutputStream states.
		// 2. Use the current buffer and internal counters to compose a `writeChunk` request.
		writeChunk = getWriteChunkRequest();
		// 3. Send the writeChunk request.
		sendWriteChunkAsync(writeChunk);
                // 4. Update putBlock data with the last writeChunk and send.
                this.containerBlockData.add(writeChunk);
                putBlockResp = executePutBlockAsync();
	}
        // 5. Wait for `putBlock` reply and issue a watchForCommmit.
	CompleteableFuture<WatchForCommitResponse> watchForCommitReps = putBlockResp.thenApply(x -> watchForCommit(x));
	// 6. Wait for `watchForCommit` response and release retries buffers accordingly (in CommitWatcher).
	WatchForCommitResponse watchForCommit = watchForCommitReps.get();
}
```

Lastly, by refactoring, I meant the required refactors to make the logic follow the order of logic (e.g. which blocks should be synchronized together and in what other, or how data is used, by passing from one step to another or getting from object context). We do not necessarily need to rebuild the current implementation.

**Comment by smengcl** [comment]
Superceded by #6859

**Pull request closed without merging**